ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -98,17 +99,49 @@ private[platform] case class StateCache[K, V](
     registerUpdateTimer,
     pendingUpdates.synchronized {
       val validAt = cacheIndex
-      val eventualValue = Future.delegate(fetchAsync(validAt))
-      val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
-      if (pendingUpdatesForKey.latestValidAt < validAt) {
-        pendingUpdatesForKey.latestValidAt = validAt
-        pendingUpdatesForKey.pendingCount += 1
-        registerEventualCacheUpdate(key, eventualValue, validAt)
-          .flatMap(_ => eventualValue)
-      } else eventualValue
+      putPendingForKey(key, validAt, Future.delegate(fetchAsync(validAt)))
     },
   )
 
+  private def putPendingForKey(key: K, validAt: Offset, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Future[V] = {
+    val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
+    if (pendingUpdatesForKey.latestValidAt < validAt) {
+      pendingUpdatesForKey.latestValidAt = validAt
+      pendingUpdatesForKey.pendingCount += 1
+      registerEventualCacheUpdate(key, eventualValue, validAt)
+        .flatMap(_ => eventualValue)
+    } else eventualValue
+  }
+
+  /** Batch fetches many contracts at once to avoid incremental fetches to database */
+  def putAsyncMany(keys: Seq[K], fetchAsync: Offset => Future[Map[K, V]], empty: K => Future[V])(
+      implicit loggingContext: LoggingContext
+  ): Future[Unit] =
+    Future
+      .sequence(
+        Timed.value(
+          registerUpdateTimer,
+          pendingUpdates.synchronized {
+            val validAt = cacheIndex
+            val eventualValue = Future.delegate(fetchAsync(validAt))
+            keys.map { key =>
+              putPendingForKey(
+                key,
+                validAt,
+                eventualValue.map(_.get(key)).transformWith {
+                  case Success(Some(value)) => Future.successful(value)
+                  case Success(None) => empty(key)
+                  case Failure(exception) => Future.failed(exception)
```

Comment:
I take back what I said before about the blowing up of the whole batch. That blow-up will resurface here and blow up all (k,v)s in that batch